packages/keystone/src/lib/core/mutations/create-update.ts

```diff
@@ -145,72 +145,73 @@ async function resolveInputForCreateOrUpdate(
 ) {
   const operation: 'create' | 'update' = existingItem === undefined ? 'create' : 'update';
   const nestedMutationState = new NestedMutationState(context);
-  let resolvedData = Object.fromEntries(
-    await promiseAllRejectWithAllErrors(
-      Object.entries(list.fields).map(async ([fieldKey, field]) => {
-        const inputConfig = field.input?.[operation];
 
-        let input = originalInput[fieldKey];
-        // Apply default values
-        if (
-          operation === 'create' &&
-          input === undefined &&
-          field.__legacy?.defaultValue !== undefined
-        ) {
-          input =
-            typeof field.__legacy.defaultValue === 'function'
-              ? await field.__legacy.defaultValue({ originalInput, context })
-              : field.__legacy.defaultValue;
-        }
+  let resolvedData = originalInput;
+
+  // Apply default values
+  // We don't expect any errors from here, so we can wrap all these operations
+  // in a generic catch-all error handler.
+  if (operation === 'create') {
+    resolvedData = Object.fromEntries(
+      await promiseAllRejectWithAllErrors(
+        Object.entries(list.fields).map(async ([fieldKey, field]) => {
+          let input = originalInput[fieldKey];
+          if (input === undefined && field.__legacy?.defaultValue !== undefined) {
+            input =
+              typeof field.__legacy.defaultValue === 'function'
+                ? await field.__legacy.defaultValue({ originalInput, context })
+                : field.__legacy.defaultValue;
+          }
+          return [fieldKey, input] as const;
+        })
+      )
+    );
+  }
 
-        // Resolve field type input resolvers
-        const resolved = inputConfig?.resolve
-          ? await inputConfig.resolve(
-              input,
-              context,
-              (() => {
-                if (field.dbField.kind !== 'relation') {
-                  return undefined;
+  // Apply field type input resolvers
+  resolvedData = Object.fromEntries(
+    await promiseAllRejectWithAllErrors(
+      Object.entries(list.fields).map(async ([fieldKey, field]) => {
+        const inputResolver = field.input?.[operation]?.resolve;
+        let input = resolvedData[fieldKey];
+        if (inputResolver) {
+          input = await inputResolver(
+            input,
+            context,
+            (() => {
+              // This third argument only applies to relationship fields
+              if (field.dbField.kind !== 'relation') {
+                return undefined;
+              }
+              if (input === undefined) {
+                // No-op: This is what we want
+                return () => undefined;
+              }
+              if (input === null) {
+                // No-op: Should this be UserInputError?
+                return () => undefined;
+              }
```

Comment:
Why is this being here better?

Reply:
It's just about visibility. I want to have a single place to point at for this discussion, and it frees up the four other resolvers to be simpler.

Reply:
Could you not do this or write it in such a way where the value used to decide what to return is based on the value passed to the function rather than the input from GraphQL? The resolve function passed to relationship field input resolvers currently accepts some input that the field's input resolver calls it with and decides what to do based on that rather than using the GraphQL input directly and this breaks that.

Reply:
I don't want to bike shed this too much because what I really want is for all this code to evaporate even further, so as long as this works I want to move forward and have these specific lines go away as part of broader changes I'm working on.